src/credentials/DeviceAttestationConstructor.cpp

```diff
@@ -0,0 +1,173 @@
+                break;
+            case kTimestampTagId:
+                VerifyOrReturnError(timestampExists == false, CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_ELEMENT);
+                ReturnErrorOnFailure(tlvReader.Get(timestamp));
+                timestampExists = true;
+                break;
+            case kFirmwareInfoTagId:
+                VerifyOrReturnError(firmwareInfoExists == false, CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_ELEMENT);
+                ReturnErrorOnFailure(tlvReader.Get(firmwareInfo));
+                firmwareInfoExists = true;
+                break;
+            default:
+                return CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_ELEMENT;
+            }
+        }
+        else if (TLV::IsProfileTag(tag))
+        {
+            // vendor fields
+            bool seenProfile = false;
+            uint16_t currentVendorId;
+            uint16_t currentProfileNum;
+
+            currentVendorId   = TLV::VendorIdFromTag(tag);
+            currentProfileNum = TLV::ProfileNumFromTag(tag);
+            if (!seenProfile)
+            {
+                seenProfile = true;
+                vendorId    = currentVendorId;
+                profileNum  = currentProfileNum;
+            }
+            else
+            {
+                // check that vendorId and profileNum match in every Vendor Reserved entry
+                VerifyOrReturnError(currentVendorId == vendorId && currentProfileNum == profileNum,
+                                    CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_ELEMENT);
+            }
+
+            ByteSpan vendorReservedEntry;
+            ReturnErrorOnFailure(tlvReader.Get(vendorReservedEntry));
+            VerifyOrReturnError(vendorReservedIdx < vendorReservedArraySize, CHIP_ERROR_NO_MEMORY);
+            vendorReservedArray[vendorReservedIdx++] = vendorReservedEntry;
+        }
+        else
+        {
+            return CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_ELEMENT;
+        }
+    }
+
+    VerifyOrReturnError(error == CHIP_END_OF_TLV, error);
+    VerifyOrReturnError(certificationDeclarationExists && attestationNonceExists && timestampExists,
+                        CHIP_ERROR_MISSING_TLV_ELEMENT);
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ConstructAttestationElements(const ByteSpan & certificationDeclaration, const ByteSpan & attestationNonce,
+                                        uint32_t timestamp, const ByteSpan & firmwareInfo, ByteSpan * vendorReservedArray,
+                                        size_t vendorReservedArraySize, uint16_t vendorId, uint16_t profileNum,
+                                        MutableByteSpan & attestationElements)
+{
+    TLV::TLVWriter tlvWriter;
+    TLV::TLVType outerContainerType = TLV::kTLVType_NotSpecified;
+
+    VerifyOrReturnError(!certificationDeclaration.empty() && !attestationNonce.empty(), CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(attestationNonce.size() == 32, CHIP_ERROR_INVALID_MESSAGE_LENGTH);
```

Comment:
This should also be CHIP_ERROR_INVALID_ARGUMENT

Reply:
will update